### ISSUE TITLE
plotMode=\bars does not draw the last element

### ISSUE THREAD

**Issue by tremblap** [supercollider/supercollider#5296]
Hello!

`(0..9).plot.plotMode=\bars;`

is unexpectedly drawing only 9 bars. My expectations might be wrong, but a 10 item array, I expected 10 bars. Boundary issues I reckon.

**Comment by prko** [comment]
\levels and \steps have the same behaviour.
https://scsynth.org/t/plot-bars-mode-problem/8683/7?u=prko

**Comment by mtmccrea** [comment]
>\levels and \steps have the same behaviour.

Yes, and my initial impression is that they should have different behavior. I think both `\levels` and `\bars` would be better suited to be centered on their domain values, which would require some special handling of the `domainSpecs` in those plots to allow for some padding at either end of the plot to accommodate.

`\steps` is ok because it follows the sample-and-hold paradigm, which seems appropriate. There could also be some consideration of padding on the ends of this plot mode, but it would have to conform to proper display of duration in the case of, e.g. plotting audio sample values.

also discussed in [#6479 ](https://github.com/supercollider/supercollider/pull/6479#issuecomment-2355188852)

**Comment by tremblap** [comment]
I think i disagree. in an oversampling case (like here, 10 values for 200 pixels) a good method might be to centre the 10 values, like a histogram. In any case, it should have 10 values equally represented in pixel.

as for SC's approach to undersampling, it is not ideal either for soundwaves. A good practice in DAWs is to take the absolute maximum distance to zero in all the samples that are 'represented' by that pixel, which is not what SC is doing, which leads so surprising drawings and commissions. other cool modes are maximum difference from the past pixel (good for disrupting waves) but hey, that would need a complete rewrite - I had that discussion many times with many SC users and I think this report is the fruit of the last one - pointing at the lowest hanging fruit.

**Comment by prko** [comment]
The following three modes display the correct number of stems:
```
(0..9).plot.plotMode=\dstems 
(0..9).plot.plotMode=\pstems
(0..9).plot.plotMode=\stems
```
It would be better to leave the bars as they are for backward compatibility.

How about creating a new mode like this?
 `\bstems`
Where b stands for 'bar'.

I think bar mode is logically correct, but it is not practical. Sometimes users need thick stems.

**Comment by mtmccrea** [comment]
Thanks for drafting an implementation of this :)

> I think bar mode is logically correct, but it is not practical. 

I disagree that current bar behavior is logically correct, because they don't align with the corresponding x (domain) values. I think your newly implemented version is logically correct, and could just be an update/fix of `\bars`. Then it could, for example, also be the default plotMode for `plotHisto`, and would conform to a more standard histogram representation.

Breaking backwards compatibility here is less of an issue I think, as it's graphics related, and I would assume it's a seldom used feature. Open to others' view on this of course.

I think a complete implementation would have to handle the issue of the first and last bars being cut off by the edge of the plot view.

**Comment by prko** [comment]
> I think a complete implementation would have to handle the issue of the first and last bars being cut off by the edge of the plot view.

Thank you for your positive feedback. I have revised my implementation and it works as you described. It also has no problem changing the plotMode by pressing m.

![Screenshot 2024-09-20 at 20 58 55](https://github.com/user-attachments/assets/9ae53c02-9759-42e5-bfff-f7fcfdc2a766)

I can make a PR with it if you like this!

**Comment by prko** [comment]
Oh, there is already #6479.
I think a new PR is not necessary. It was a good exercise for me.

**Comment by mtmccrea** [comment]
> Oh, there is already #6479.
I think a new PR is not necessary. 

@elgiano did mention they may not have time to continue that PR in the near term, so perhaps you could combine efforts on this if you'd like...

**Comment by prko** [comment]
@mtmccrea 

OK! 
After reading the thread #6497, I have a question:

The first implementation seems simple and logical to me.
The labels and grids of the second one should be corrected as follows:

![369393334-9ae53c02-9759-42e5-bfff-f7fcfdc2a766](https://github.com/user-attachments/assets/9bb9500d-eb78-408c-92ff-3af8c6a88ef4)

However, there is the problem of misalignment of x-lables and x-grids when switching modes: 
![image](https://github.com/user-attachments/assets/5322bee1-3cac-455f-89de-c86271dab613)


The first method could be done immediately as a PR, but I should think more about implementing the third one.

Which one of the three ways is the best for you?

**Comment by mtmccrea** [comment]
> Which one of the three ways is the best for you?

It's not clear which three you're referring to.

I'd imagine the grids/labels don't need to be altered (the domain spec can extend the bounds as needed), just the offset of the drawing of the bars should change.

Would also be interesting to test the `editable` mode to make sure it behaves properly. Then these can serve as faders :)

**Comment by tremblap** [comment]
Indeed the top one is right for bars, and the bottom right one for samples. Thanks for sorting this out!

**Comment by prko** [comment]
Moving x values is not easy...

Anyway, I think the best way is to 
- move the whole x-axis to the right 
- separate two axes

as follows:

![Image](https://github.com/user-attachments/assets/f22f1663-2d17-4488-8396-7b4f5a0f645f)

This is the leftmost value and the rightmost value appears in its full form.

**Comment by mtmccrea** [comment]
Centering `\bars` on their x-axis index position would already be a useful improvement (as in your bottom center bars example). Then the issue of padding the x dimension on both sides (e.g. by a half-step) could be implemented by internally setting the `domain` and `domainSpecs`, for example:
```supercollider
p = (0..9).plot.plotMode_(\points);
p.domain_((0..9)).domainSpecs_([-0.5,9.5].asSpec)
```
There's a question of whether this should only be applied to `\bars` plot mode or to other modes as well (which would take more testing of various use cases).
